### DISCUSSION TITLE
async function as input for x-data

### DISCUSSION THREAD

**Discussion by matthiasstorr** [alpinejs/alpine#2287] [1. Help] [answered]
Hello, 

i am currently trying to get my code snippet to work but at the moment i have no idea to proceed further - maybe one of you can lead me in the right direction:

I want to fetch data from an api asynchronously. I get it to work by the following code snippet. 
**Working example:**

```javascript
<div x-data="devices()" 
         x-init="foo.getMyDevices()
                     .then(data => {myDevices = {...data.devices};})
                     .catch(err => console.log(err));">

<script> function devices() {
               return {
                            myDevices: {}
              }
            };
</script>
```

But at best it should look something like this. But then i get the error, that _Object.defineProperty is called on non-object_
**How i actually want it to work:**

```javascript
<div x-data="getMyDevices()">
<script> const getMyDevices = async () => {
                 const response = await foo.getMyDevices();
                 const myDevices = await {...response.devices};
                 return {
                    myDevices
                 }
              };
</script>
```

**Comment by SimoTod** [comment, marked as answer]
An async function returns a promise which breaks alpine since it needs a standard  object for the scope. I never tried if it works for x-data but you can try to use `await` in the directive: https://alpinejs.dev/advanced/async

**Reply by matthiasstorr** [reply]
Thanks for your fast reply. I actually tried it but didn’t work. So I will use my workaround.

**Comment by SimoTod** [comment]
Yeah, I wasn't 100% sure because x-data does other stuff as well. Maybe
it's not possible at the moment
